immutablecollections/_immutabledict.py

```diff
@@ -146,6 +146,9 @@ def __repr__(self):
     def __str__(self):
         return "{%s}" % ", ".join(["%s: %s" % item for item in self.items()])
 
+    def __reduce__(self):
+        return (immutabledict, (list(self.items()),))
```

Comment:
Why `list` and not `tuple`?

Reply:
hmm - this will lose the "singleton-ness" of empty `ImmutableDict`s in the serialized form (though it will be restored at de-serialization time).  Can you do a quick test to see if the size savings is significant if you have a large number of them?

Reply:
1. No reason for `list`, I'll change it.
2. Like this? (size is in bytes)
```python
In [25]: from sys import getsizeof

In [26]: d = dict()

In [27]: d
Out[27]: {}

In [28]: list(d.items())
Out[28]: []

In [29]: getsizeof(list(d.items()))
Out[29]: 64

In [30]: getsizeof(())
Out[30]: 48

In [31]: getsizeof(tuple(d.items()))
Out[31]: 48
```

Reply:
Maybe you meant timing, this is interesting:
```python
In [33]: %timeit tuple(d.items())
197 ns ± 3.31 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [34]: %timeit x = tuple(d.items())
193 ns ± 3.18 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [35]: %%timeit
    ...: if not d:
    ...:     x = ()
    ...: else:
    ...:     x = tuple(d.items())
    ...:
23.3 ns ± 0.834 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Reply:
Additional timings:
```python
In [37]: %timeit x = tuple(d.items()) if len(d) else ()
53.6 ns ± 1.79 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [38]: %timeit x = tuple(d.items()) if d else ()
24.7 ns ± 1.57 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [39]: %timeit x = () if not d else tuple(d.items())
24.8 ns ± 1.28 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Reply:
@berquist : what I was after is the size of the serialized form of the representation of an empty `ImmutableDict` vs. a Pickle "backreference to a previously serialized `object`.  You could get this by measuring the number of bytes in the pickled form of, let's say, a tuple 10k empty `immutabledicts` with each implementation (return a new empty tuple vs a constant singleton empty tuple; 10k because there is probably some pickle header overhead).  But it occurs to me that it shouldn't matter at least on CPython because I believe the empty tuple is always a singleton anyway.

Reply:
Empty collections are pretty common at least in our code so the time optimizations to check for empty ones may be worthwhile

Reply:
Correct on the last point. Here are timings for the two approaches on 10k `immutabledict`s:
before:
```python
In [12]: %timeit [dumps(d) for d in dicts]
321 ms ± 14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
after:
``` python
In [7]: %timeit [dumps(d) for d in dicts]
195 ms ± 13.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
And the size of 10k empty tuples is the same as 10k empty `immutabledict`s, both before and after pickling.

Reply:
Can you do timings where 80% of the dicts are small but non-empty and 20% are empty? We want to make sure the special case does not sacrifice too much in the non-empty case.

When you say the size is the same before and after pickling, do you mean of the in-memory Python objects or also the serialized bytes? (I expect it to be the same in both cases)

Reply:
IPython timing code:
```python
from immutablecollections import immutabledict
import random
from pickle import dumps

empty_tuple = ()
empty_list = []
empty_dict = {}

small_tuple = (3, 1, 2)
small_list = [3, 1, 2]
small_dict = {"a": 3, "b": 1, "c": 2}

rand = random.Random(0)

big_dim = int(1e4)
large_part = int(0.8 * big_dim)
small_part = big_dim - large_part

big_list = list(range(big_dim))
rand.shuffle(big_list)
big_tuple = tuple(big_list)
rand.shuffle(big_list)
big_dict = dict(zip(big_list, big_tuple))

empty_immutabledict = immutabledict()
small_immutabledict = immutabledict(small_dict)
big_immutabledict = immutabledict(big_dict)

empty_dicts = [empty_immutabledict for _ in range(big_dim)]
small_dicts = [small_immutabledict for _ in range(big_dim)]
big_dicts = [big_immutabledict for _ in range(big_dim)]

mixed_dicts = []
for _ in range(large_part):
    mixed_dicts.append(small_immutabledict)
for _ in range(small_part):
    mixed_dicts.append(empty_immutabledict)
rand.shuffle(mixed_dicts)

print("empty")
%timeit [dumps(d) for d in empty_dicts]
print("mixed")
%timeit [dumps(d) for d in mixed_dicts]
print("small")
%timeit [dumps(d) for d in small_dicts]
print("big")
%timeit [dumps(d) for d in big_dicts]
```
Here are results for the generic implementation using lists:
```
empty
31.8 ms ± 2.34 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
mixed
44.9 ms ± 2.97 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
small
48.4 ms ± 3.35 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
big
45.5 s ± 417 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Here are results for the generic implementation using tuples:
```
empty
30.2 ms ± 1.55 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
mixed
41.9 ms ± 2.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
small
44.8 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
big
45.8 s ± 690 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Here are results for the special cased implementation using tuples:
```
empty
20 ms ± 267 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
mixed
44.8 ms ± 1.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
small
52.4 ms ± 1.71 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
big
47.3 s ± 849 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I also wrote a benchmark with full results [here](https://gist.githubusercontent.com/berquist/ec3b46735b3059402b1c3a87a3fcf596/raw). Here are only the `immutabledict` results, and the key is
- `0001_as_list`: serialize to a list
- `0002_as_tuple`: serialize to a tuple
- `0003_as_tuple_special_case`: serialize to a tuple, giving special preference to the empty tuple

ordered in terms of speed:
```
-------------------------------------------------------------------------------------------------------------- benchmark 'Serialization': 54 tests ---------------------------------------------------------------------------------------------------------------
Name (time in ns)                                  Min                         Max                       Mean                     StdDev                     Median                        IQR              Outliers             OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
empty immutabledict (0003_as_tupl)          3,000.0000 (9.33)          73,000.0000 (24.33)         4,130.2814 (9.90)          1,971.6079 (16.18)         4,000.0000 (10.18)         1,000.0000 (23.00)      927;1464    242,114.2543 (0.10)      35715           1
empty immutabledict (0001_as_list)          4,000.0000 (12.44)        116,000.0000 (38.67)         5,683.3238 (13.63)         2,419.9994 (19.86)         5,000.0000 (12.73)         1,000.0000 (23.00)      745;1418    175,953.3743 (0.07)      38462           1
empty immutabledict (0002_as_tupl)          4,000.0000 (12.44)         66,000.0000 (22.00)         5,808.6638 (13.93)         2,331.1851 (19.13)         5,000.0000 (12.73)         1,000.0000 (23.00)     1066;1479    172,156.6319 (0.07)      33334           1
small immutabledict (0001_as_list)          6,000.0000 (18.67)         99,000.0000 (33.00)         7,406.5051 (17.76)         2,841.3192 (23.32)         7,000.0000 (17.82)         2,000.0000 (46.00)     1236;1236    135,016.4462 (0.06)      38462           1
small immutabledict (0002_as_tupl)          6,000.0000 (18.67)        107,000.0000 (35.67)         7,427.7914 (17.81)         2,604.1426 (21.37)         7,000.0000 (17.82)         1,000.0000 (23.00)       551;617    134,629.5204 (0.06)      33334           1
small immutabledict (0003_as_tupl)          6,000.0000 (18.67)         91,000.0000 (30.33)         8,101.1328 (19.43)         3,177.0271 (26.07)         7,000.0000 (17.82)         1,000.0000 (23.00)     1103;2353    123,439.5274 (0.05)      20834           1
big immutabledict (0001_as_list)       63,888,000.0000 (>1000.0)   93,477,000.0000 (>1000.0)  72,440,500.0000 (>1000.0)   8,989,737.1640 (>1000.0)  68,865,000.0000 (>1000.0)   4,390,000.0000 (>1000.0)         2;2         13.8044 (0.00)         10           1
big immutabledict (0003_as_tupl)       65,018,000.0000 (>1000.0)  126,291,000.0000 (>1000.0)  79,905,000.0000 (>1000.0)  20,266,743.6087 (>1000.0)  68,375,000.0000 (>1000.0)  18,380,000.0000 (>1000.0)         1;1         12.5149 (0.00)          9           1
big immutabledict (0002_as_tupl)       65,624,000.0000 (>1000.0)   91,055,000.0000 (>1000.0)  71,650,666.6667 (>1000.0)   8,429,522.5250 (>1000.0)  68,068,000.0000 (>1000.0)   7,392,000.0000 (>1000.0)         1;1         13.9566 (0.00)          9           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
ordered in terms of 1, 2, 3:
```
-------------------------------------------------------------------------------------------------------------- benchmark 'Serialization': 54 tests ---------------------------------------------------------------------------------------------------------------
Name (time in ns)                                  Min                         Max                       Mean                     StdDev                     Median                        IQR              Outliers             OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
empty immutabledict (0001_as_list)          4,000.0000 (12.44)        116,000.0000 (38.67)         5,683.3238 (13.63)         2,419.9994 (19.86)         5,000.0000 (12.73)         1,000.0000 (23.00)      745;1418    175,953.3743 (0.07)      38462           1
empty immutabledict (0002_as_tupl)          4,000.0000 (12.44)         66,000.0000 (22.00)         5,808.6638 (13.93)         2,331.1851 (19.13)         5,000.0000 (12.73)         1,000.0000 (23.00)     1066;1479    172,156.6319 (0.07)      33334           1
empty immutabledict (0003_as_tupl)          3,000.0000 (9.33)          73,000.0000 (24.33)         4,130.2814 (9.90)          1,971.6079 (16.18)         4,000.0000 (10.18)         1,000.0000 (23.00)      927;1464    242,114.2543 (0.10)      35715           1
small immutabledict (0001_as_list)          6,000.0000 (18.67)         99,000.0000 (33.00)         7,406.5051 (17.76)         2,841.3192 (23.32)         7,000.0000 (17.82)         2,000.0000 (46.00)     1236;1236    135,016.4462 (0.06)      38462           1
small immutabledict (0002_as_tupl)          6,000.0000 (18.67)        107,000.0000 (35.67)         7,427.7914 (17.81)         2,604.1426 (21.37)         7,000.0000 (17.82)         1,000.0000 (23.00)       551;617    134,629.5204 (0.06)      33334           1
small immutabledict (0003_as_tupl)          6,000.0000 (18.67)         91,000.0000 (30.33)         8,101.1328 (19.43)         3,177.0271 (26.07)         7,000.0000 (17.82)         1,000.0000 (23.00)     1103;2353    123,439.5274 (0.05)      20834           1
big immutabledict (0001_as_list)       63,888,000.0000 (>1000.0)   93,477,000.0000 (>1000.0)  72,440,500.0000 (>1000.0)   8,989,737.1640 (>1000.0)  68,865,000.0000 (>1000.0)   4,390,000.0000 (>1000.0)         2;2         13.8044 (0.00)         10           1
big immutabledict (0002_as_tupl)       65,624,000.0000 (>1000.0)   91,055,000.0000 (>1000.0)  71,650,666.6667 (>1000.0)   8,429,522.5250 (>1000.0)  68,068,000.0000 (>1000.0)   7,392,000.0000 (>1000.0)         1;1         13.9566 (0.00)          9           1
big immutabledict (0003_as_tupl)       65,018,000.0000 (>1000.0)  126,291,000.0000 (>1000.0)  79,905,000.0000 (>1000.0)  20,266,743.6087 (>1000.0)  68,375,000.0000 (>1000.0)  18,380,000.0000 (>1000.0)         1;1         12.5149 (0.00)          9           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

> When you say the size is the same before and after pickling, do you mean of the in-memory Python objects or also the serialized bytes? (I expect it to be the same in both cases)

Here is using tuples:
``` python
serialized_empty_dicts = [dumps(d) for d in empty_dicts]
serialized_mixed_dicts = [dumps(d) for d in mixed_dicts]
serialized_small_dicts = [dumps(d) for d in small_dicts]
serialized_big_dicts = [dumps(d) for d in big_dicts]

deserialized_empty_dicts = [loads(d) for d in serialized_empty_dicts]
deserialized_mixed_dicts = [loads(d) for d in serialized_mixed_dicts]
deserialized_small_dicts = [loads(d) for d in serialized_small_dicts]
deserialized_big_dicts = [loads(d) for d in serialized_big_dicts]

from sys import getsizeof

print(getsizeof(empty_dicts))
print(getsizeof(mixed_dicts))
print(getsizeof(small_dicts))
print(getsizeof(big_dicts))

print(sum([getsizeof(d) for d in empty_dicts]))
print(sum([getsizeof(d) for d in mixed_dicts]))
print(sum([getsizeof(d) for d in small_dicts]))
print(sum([getsizeof(d) for d in big_dicts]))

print(sum([len(d) for d in serialized_empty_dicts]))
print(sum([len(d) for d in serialized_mixed_dicts]))
print(sum([len(d) for d in serialized_small_dicts]))
print(sum([len(d) for d in serialized_big_dicts]))

print(sum([getsizeof(d) for d in serialized_empty_dicts]))
print(sum([getsizeof(d) for d in serialized_mixed_dicts]))
print(sum([getsizeof(d) for d in serialized_small_dicts]))
print(sum([getsizeof(d) for d in serialized_big_dicts]))

print(sum([getsizeof(d) for d in deserialized_empty_dicts]))
print(sum([getsizeof(d) for d in deserialized_mixed_dicts]))
print(sum([getsizeof(d) for d in deserialized_small_dicts]))
print(sum([getsizeof(d) for d in deserialized_big_dicts]))
```
```
87624
87624
87624
87624

480000
480000
480000
480000

630000
958000
1040000
1187980000

960000
1288000
1370000
1188310000

480000
480000
480000
480000
```
I think `getsizeof` isn't working as expected because it isn't recursively checking the size of all objects inside compound structures, it's only looking at the top level, so it's returning the pointer size (as evidenced by the all sizes in the first/second/fourth groups being the same). So what I said about sizes isn't right.

Reply:
@berquist : Are your speed charts using pytest-benchmark? Looking at the median and averages, it looks like if we special-case for empty we don't lose anything significant in non-empty cases and gain a bit for the empties, so we might as well do it.

For size, `getsizeof` does only look at top-level - sorry,  should have highlighted that problem earlier.

Reply:
Yes, those speed charts are with pytest-benchmark. Should I add it to this PR?

Reply:
Sure, please do

Reply:
(just the speed parts)